l3kernel/l3tl.dtx

```diff
@@ -3087,36 +3162,50 @@
 % \end{macro}
 %
 % \begin{macro}[pTF]{\tl_if_head_is_N_type:n}
-% \begin{macro}[EXP]{\@@_if_head_is_N_type:w}
+% \begin{macro}[EXP]
+%   {
+%     \@@_if_head_is_N_type_auxi:w   ,
+%     \@@_if_head_is_N_type_auxii:nn ,
+%     \@@_if_head_is_N_type_auxiii:n
+%   }
 %   A token list can be empty, can start with an explicit space
 %   character (catcode 10 and charcode 32), can start with a begin-group
 %   token (catcode 1), or start with an \texttt{N}-type argument.  In
-%   the first two cases, the line involving \cs{@@_if_head_is_N_type:w}
-%   produces~|^| (and otherwise nothing).  In the third case
-%   (begin-group token), the lines involving \cs{exp_after:wN} produce a
+%   the first two cases, the line involving \cs{@@_if_head_is_N_type_auxi:w}
+%   produces~|f| (and otherwise nothing).  In the third case
+%   (begin-group token), the lines involving \cs{token_to_str:N} produce a
 %   single closing brace.  The category code test is thus true exactly
 %   in the fourth case, which is what we want.  One cannot optimize by
-%   moving one of the |*| to the beginning: if |#1| contains primitive
+%   moving one of the |\scan_stop:| to the beginning: if |#1| contains primitive
 %   conditionals, all of its occurrences must be dealt with before the
 %   \cs{if_catcode:w} tries to skip the \texttt{true} branch of the
 %   conditional.
 %    \begin{macrocode}
 \prg_new_conditional:Npnn \tl_if_head_is_N_type:n #1 { p , T , F , TF }
   {
-    \if_catcode:w
-        \if_false: { \fi: \@@_if_head_is_N_type:w ? #1 ~ }
-        \exp_after:wN \use_none:n
-          \exp_after:wN { \exp_after:wN { \token_to_str:N #1 ? } }
-        * *
+    \if:w
+        \if_false: { \fi: \@@_if_head_is_N_type_auxi:w \prg_do_nothing: #1 ~ }
+        { \exp_after:wN { \token_to_str:N #1 } }
+        \scan_stop: \scan_stop:
       \prg_return_true:
     \else:
       \prg_return_false:
     \fi:
   }
-\cs_new:Npn \@@_if_head_is_N_type:w #1 ~
+\cs_new:Npn \@@_if_head_is_N_type_auxi:w #1 ~
   {
-    \tl_if_empty:oTF { \use_none:n #1 } { ^ } { }
-    \exp_after:wN \use_none:n \exp_after:wN { \if_false: } \fi:
+    \tl_if_empty:oTF { #1 }
+      { \exp_after:wN \@@_if_head_is_N_type_auxii:nn }
+      { \exp_after:wN \@@_if_head_is_N_type_auxiii:n }
+    \exp_after:wN { \if_false: } \fi: 
+  }
+\cs_new:Npn \@@_if_head_is_N_type_auxii:nn #1 #2
+  {
+    f
+  }
+\cs_new:Npn \@@_if_head_is_N_type_auxiii:n #1
+  {
+    \exp_after:wN \use_none:n \exp_after:wN 
   }
```

Comment:
Perhaps replace `\exp_after:wN \@@_if_head_is_N_type_auxii:nn` by `f \exp_after:wN \use_none:nn`.
Perhaps put the code of `\@@_if_head_is_N_type_auxiii:n` in one line (and rename to `auxii`).